services/src/main/scala/cromwell/services/instrumentation/AsynchronousThrottlingGaugeMetricActor.scala

```diff
@@ -0,0 +1,73 @@
+package cromwell.services.instrumentation
+
+import akka.actor.{ActorRef, LoggingFSM, Props, Status}
+import akka.pattern.pipe
+import cats.data.NonEmptyList
+import cromwell.services.instrumentation.AsynchronousThrottlingGaugeMetricActor._
+
+import scala.concurrent.Future
+
+class AsynchronousThrottlingGaugeMetricActor(metricPath: NonEmptyList[String],
+                                             instrumentationPrefix: Option[String],
+                                             override val serviceRegistryActor: ActorRef)
+  extends LoggingFSM[AsynchronousThrottlingGaugeMetricActorState, Unit]
+    with CromwellInstrumentation {
+
+  implicit val ec = context.dispatcher
+
+  startWith(WaitingForMetricCalculationRequestOrMetricValue, ())
+
+  when(WaitingForMetricCalculationRequestOrMetricValue) {
+    case Event(MetricValue(value), _) =>
+      sendGaugeAndStay(value)
+    case Event(CalculateMetricValue(calculateMetricValueFunction), _) =>
+      calculateMetricValueFunction()
+        .map(intValue => FinishedMetricValueCalculation(intValue.toLong))
+        .pipeTo(self)
+      goto(MetricCalculationInProgress)
+  }
+
+  when(MetricCalculationInProgress) {
+    case Event(CalculateMetricValue, _) =>
+      // metric actor is already busy calculating metric value, so we dismiss this request
+      stay()
+    case Event(MetricValue(value), _) =>
+      // pass through the pre-calculated metric value and stay, continuing to wait for ongoing metric value calculation to finish
+      sendGaugeAndStay(value)
+    case Event(FinishedMetricValueCalculation(value), _) =>
+      sendGauge(metricPath, value, instrumentationPrefix)
+      goto(WaitingForMetricCalculationRequestOrMetricValue)
+    case Event(Status.Failure(ex), _) =>
+      log.error(s"Cannot send gauge metric value: error occurred while evaluating Future value.", ex)
+      goto(WaitingForMetricCalculationRequestOrMetricValue)
+  }
+
+  whenUnhandled {
+    case Event(unexpected, _) =>
+      log.warning(s"Programmer error: this actor should not receive message $unexpected while in state $stateName")
+      stay()
+  }
+
+  private def sendGaugeAndStay(metricValue: Long): State = {
+    sendGauge(metricPath, metricValue, instrumentationPrefix)
+    stay()
+  }
+
+}
+
+object AsynchronousThrottlingGaugeMetricActor {
+
+  def props(metricPath: NonEmptyList[String], instrumentationPrefix: Option[String], serviceRegistryActor: ActorRef) =
+    Props(new AsynchronousThrottlingGaugeMetricActor(metricPath, instrumentationPrefix, serviceRegistryActor))
+
+  // states
+  sealed trait AsynchronousThrottlingGaugeMetricActorState
+  case object WaitingForMetricCalculationRequestOrMetricValue extends AsynchronousThrottlingGaugeMetricActorState
+  case object MetricCalculationInProgress extends AsynchronousThrottlingGaugeMetricActorState
+
+  // messages
+  case class CalculateMetricValue(calculateMetricValueFunction: () => Future[Int])
```

Comment:
Double check me on this but I think you want to take an `implicit ec: ExecutionContext` for this function. Otherwise I think the ec used will be from the actor which originally created the function, rather than the actor which is running it.

Reply:
I tried, but looks like there is no way to do that.

> Otherwise I think the ec used will be from the actor which originally created the function

That seems true, but I'm not sure that this is exactly bad.

Reply:
Something like this? https://github.com/broadinstitute/cromwell/commit/33e2f7d87782abdef2dba432b8d852b3d63b46af

In general, I think the ec should be supplied as late as possible, and (to me at least!) in akka that means the actor which is doing the work should be the one providing the execution context. 

Ultimately, in this case is probably won't make too much practical difference because the threads will all be coming from the services thread pool anyway, but I think it's a good habit to be in. And if the threads are named based on the actor which they come from, it might make debugging a little easier too.

Reply:
That works, thank you. Earlier I was trying to make it implicit somehow and didn't succeed